src/code/irqmgr.c

```diff
@@ -1,122 +1,179 @@
+ *      happening at approximately 60Hz on NTSC and 50Hz on PAL. Many threads sit idle
+ *      until a VI Retrace event wakes them up, at which point they will perform their
+ *      task and then return to idle to await the next retrace.
+ *
+ *  - Pre-NMI
+ *      This event is sent to the IRQ manager by the OS Interrupt Handler when the reset
+ *      button on the N64 control deck is pressed. This event is forwarded to clients to
+ *      inform them that a reset will occur in at least 0.5s / 500ms so they may begin any
+ *      shutdown procedures.
+ *
+ *  - NMI
+ *      This event is sent at 450ms into the Pre-NMI phase, informing clients that the
+ *      mandatory 0.5s of Pre-NMI is almost over and a reset may be imminent. This event
+ *      is not to be confused with the hardware NMI interrupt signalled when the CPU is
+ *      to fully reset, as by the time that interrupt is received there is no time left
+ *      to do anything.
+ */
 #include "global.h"
 #include "vt.h"
 
-vu32 gIrqMgrResetStatus = 0;
+vu32 gIrqMgrResetStatus = IRQ_RESET_STATUS_IDLE;
 volatile OSTime sIrqMgrResetTime = 0;
 volatile OSTime gIrqMgrRetraceTime = 0;
 u32 sIrqMgrRetraceCount = 0;
 
-#define RETRACE_MSG 666
-#define PRE_NMI_MSG 669
-#define PRENMI450_MSG 671
-#define PRENMI480_MSG 672
-#define PRENMI500_MSG 673
-
-#define STATUS_IDLE 0
-#define STATUS_PRENMI 1
-#define STATUS_NMI 2
-
-void IrqMgr_AddClient(IrqMgr* this, IrqMgrClient* c, OSMesgQueue* msgQ) {
+// Internal messages
+#define IRQ_RETRACE_MSG 666
+#define IRQ_PRENMI_MSG 669
+#define IRQ_PRENMI450_MSG 671
+#define IRQ_PRENMI480_MSG 672
+#define IRQ_PRENMI500_MSG 673
+
+/**
+ * Registers a client and an associated message queue with the IRQ manager. When an
+ * interrupt event is received by the IRQ manager, these clients will be notified of
+ * the event.
+ *
+ * @param irqMgr the IrqMgr instance to register with.
+ * @param client client to register.
+ * @param msgQ message queue to send notifications of interrupts to, associated with the client.
+ */
+void IrqMgr_AddClient(IrqMgr* irqMgr, IrqMgrClient* client, OSMesgQueue* msgQ) {
     OSIntMask prevInt;
 
-    LogUtils_CheckNullPointer("this", this, "../irqmgr.c", 96);
-    LogUtils_CheckNullPointer("c", c, "../irqmgr.c", 97);
+    LogUtils_CheckNullPointer("this", irqMgr, "../irqmgr.c", 96);
+    LogUtils_CheckNullPointer("c", client, "../irqmgr.c", 97);
     LogUtils_CheckNullPointer("msgQ", msgQ, "../irqmgr.c", 98);
 
     prevInt = osSetIntMask(OS_IM_NONE);
 
-    c->queue = msgQ;
-    c->prev = this->clients;
-    this->clients = c;
+    client->queue = msgQ;
+    client->prev = irqMgr->clients;
+    irqMgr->clients = client;
 
     osSetIntMask(prevInt);
 
-    if (this->resetStatus > STATUS_IDLE) {
-        osSendMesg(c->queue, (OSMesg) & this->prenmiMsg, OS_MESG_NOBLOCK);
+    if (irqMgr->resetStatus >= IRQ_RESET_STATUS_PRENMI) {
+        osSendMesg(client->queue, (OSMesg)&irqMgr->prenmiMsg, OS_MESG_NOBLOCK);
```

Comment:
Does it make sense to cast pointers to `OSMesg` since that's just a typedef for `void*`? I'm psure it wouldn't cause any warnings at least, but are you supposed to cast to OSMesg regardless of the type of the message or something like that?

Also noticed we have a lot of incorrect casts for osRecvMesg calls which cast to `OSMesg` even tho it takes a `OSMesg*`, but that's another issue

Reply:
You're correct it doesn't cause any warnings if the cast were omitted. It's mostly for clarity I believe? There's no necessary convention that you should cast to OSMesg afaik. If you think it's better off implicit then I can change that.

Reply:
I'm not sure tbh, and thinking about it more it's really a general thing so we don't have to change it with this PR

Reply:
A separate PR sounds good, I have some things to clean up with message queues and threads in general in a separate PR already so we can probably deal with this more generally there.